### DISCUSSION TITLE
My Experience after a 2 day project. Some wishes and questions

### DISCUSSION THREAD

**Discussion by adakordev** [twinbasic/twinbasic#507] [Ideas]
Over the last 2 days I have created a small project with twinBasic and here are some points I would like to mention.
But first, thank you for your work, it already looks great. 

1.)
Why is this possible and is not caught early, is it possible to fix this?

Dim bool As Boolean = "test"
Dim i As Integer = 2 > 4	

By accident I had assigned a statement/ Boolean in my Integer variable, which came out as 0 during runtime and caused problems.

2.)
So I wrote and hit Enter, for example
For rowCounter = 44 To 5555

This is what I got. The “Next” is not indented the same way as the “For”
Is it possible to fix this and automatically add “rowCounter” to “Next” so I receive “Next rowCounter” after I hit Enter?
 

![forNext](https://user-images.githubusercontent.com/31490758/143623861-fc451561-dcec-4832-8dfb-93a12f77f73e.JPG)
3.)
I have a GUI and after pressing a Button a Filedialog (xlApp.Application.FileDialog(msoFileDialogFilePicker) opens up. Is it possible to open the Filedialog a layer above the GUI? Right now it is behind the GUI after the Filedialog opens.

4.)
Is it possible to add a custom icon to the .exe file?

5.)
I guess additional GUI controls like a dropdown or the possibility to add pictures to the .tbform will come at a later date.

Thank you very much!

Windows 10
tB 0.13.4

**Comment by WaynePhillipsEA** [comment]
Thanks for your feedback!

> 1.) Why is this possible and is not caught early, is it possible to fix this?
> Dim bool As Boolean = "test" 
> Dim i As Integer = 2 > 4

This works correctly as per VBx... with the first line producing a runtime error and the second being allowed since a Boolean can be assigned to an integer via implicit coersion.    

It is planned that these types of problems will be caught using the analysis extension API #51, rather than in the compiler itself.  Additionally we also want to offer Option Strict as seen in VB.NET that would disallow the implicit coersions altogether, see #33 .

> 2.) So I wrote and hit Enter, for example For rowCounter = 44 To 5555
> 
> This is what I got. The “Next” is not indented the same way as the “For” Is it possible to fix this and automatically add “rowCounter” to “Next” so I receive “Next rowCounter” after I hit Enter?

Yes, that's a bug.  I'll have a look at that.

> I have a GUI and after pressing a Button a Filedialog (xlApp.Application.FileDialog(msoFileDialogFilePicker) opens up. Is it possible to open the Filedialog a layer above the GUI? Right now it is behind the GUI after the Filedialog opens.

Excel runs in a separate process, so you'd be better using a native dialog that runs in-process.  Something like the old Ken Getz one (https://gist.github.com/WaynePhillipsEA/91f2dc59f06a632f848a9bb5c66d4a17) would work.   You could tweak it to get the current form HWND by using GetForegroundWindow, and using that for the parent HWND.

> 4.) Is it possible to add a custom icon to the .exe file?

Yes.  If you look in the file explorer in VS code, you'll find a 'Resources' folder in your project.  Inside that you'll see an ICON folder with an icon called twinBASIC.ico.   You can delete that file, and put your own ICO file instead.

> 5.) I guess additional GUI controls like a dropdown or the possibility to add pictures to the .tbform will come at a later date.

Yup, more GUI work coming in December.

**Comment by WaynePhillipsEA** [comment]
> This is what I got. The “Next” is not indented the same way as the “For”

I've tried to reproduce this, but the 'Next' always seems to get correctly placed for me.

**Comment by adakordev** [comment]
Well, I tried to reproduce the 'For' 'Next' situation in a new project, but was not able to yet. 
It just does not work in this one project I mentioned. I will take a look again later.

> Yes. If you look in the file explorer in VS code, you'll find a 'Resources' folder in your project. Inside that you'll see an ICON folder with an icon called twinBASIC.ico. You can delete that file, and put your own ICO file instead.

When I open my project, which was created using your Sample 6 CustomControls, I see the "Resources" folder, yes. But when I click this folder to expand, it will be renamed to "Resources / BITMAP" every time. There is no ICON folder in there. When I try to create an ICON folder I receive following messages:

An unknown error occurred. Please consult the log for more details.
folder name must be in the format LCID_0000, e.g. LCID_0409 for US English

![image](https://user-images.githubusercontent.com/31490758/143679621-33066fa1-1964-4172-9364-5b39c930d6f8.png)

I created a fresh new Sample 6 CustomControls and it was the same.


After creating a new Project with different template "Create Sample 1. HelloWorld" I can see the Resources/ ICON folder and the twinBASIC.ico in there. When I replace the .ico with my own, it does not work and it just stays the default icon after building the .exe. Is a certain size required?

![image](https://user-images.githubusercontent.com/31490758/143680008-8355ba43-a296-4b00-8a22-c2ce00fd0247.png)

**Reply by WaynePhillipsEA** [reply]
You'll want to turn off the 'Compact Folders' setting in VS Code, as that is why the 'BITMAP' folder is appearing like that in your screenshot, and that is confusing things.   File menu > Preferences > Settings > search for 'Compact Folders', and turn it off.

Windows caches EXE icons, and so if you change the icon in your build then you'll have to reset the icon cache in Windows, otherwise you'll get the stale icon in Windows Explorer.   You can do this easily with the command "ie4uinit.exe -show" run from the command line.   Immediately after running that command, you should find that the icon is shown correctly in Explorer.

**Comment by adakordev** [comment]
It seems I have found the reason for the false 'for Next' indentation.
Somewhere above I had the following 2 lines and yes, they were/ are a comment:

**'MsgBox "Die Erstellung der Datei dauert in dieser TEST - Version noch etwas." & vbLf & vbLf & _
"Wir bitten um Geduld. Der weiße screen während der Wartezeit ist in Ordnung.", vbInformation, "Bitte beachten"**

It seems this was the reason and not only the For loops, but also With blocks were affected and indented false.

I would love to hear any idea on my previous post with the .ico in the coming days as well. Thank you for your time.
